### DISCUSSION TITLE
git push issue

### DISCUSSION THREAD

**Discussion by isacall099** [cafawo/FinancialDataAnalytics#5] [Q&A] [answered]
Hi, I pushed changes to my repository several times, mostly they were small changes adding notes to the script and etc.  In last push I received this error.
! [rejected]        main -> main (non-fast-forward)
error: failed to push some refs to 'https://github.com/isacall099/220213600'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
As part of the solution, I tried to pull, but after " add  . " I received many other messages, to not destroy the repository somehow, I stopped trying. Overall, I am unable to push local repository to remote one.
I received the following after using git pull
hint: You have divergent branches and need to specify how to reconcile them.
hint: You can do so by running one of the following commands sometime before
hint: your next pull:
hint: 
hint:   git config pull.rebase false  # merge (the default strategy)
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint: 
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.

**Comment by birte0** [comment, marked as answer]
The error message indicates that your local branch is behind the remote branch, and you need to integrate the remote changes before pushing again. 

To resolve this issue, you can try pulling the latest changes from the remote repository using the "git pull" command. However, it seems like you encountered some other messages after running "git pull", which suggest that you have divergent branches and need to specify how to reconcile them. You can do this by running one of the following commands before your next pull:

- `git config pull.rebase false` (merge, the default strategy)
- `git config pull.rebase true` (rebase)
- `git config pull.ff only` (fast-forward only)

You can replace "git config" with "git config --global" to set a default preference for all repositories. Additionally, you can pass --rebase, --no-rebase, or --ff-only on the command line to override the configured default per invocation.

Once you've specified how to reconcile the divergent branches, you should be able to pull the remote changes and resolve the conflict. After resolving the conflict, you can push your local changes to the remote repository.

Alternatively, if you're sure that your local branch is the correct one, you can use "git push --force" to push your local changes to the remote repository, even if they conflict with the remote changes. However, this is considered a brute force solution and should be used with caution.

I hope this helps! Let me know if you have any further questions.

**Reply by cafawo** [reply]
@isacall099 please up-vote and mark this as the answer if it helped to resolve your issue.

**Reply by isacall099** [reply]
Thank you it worked
